storage/filestore/dir.go

```diff
@@ -273,48 +265,61 @@ func (dir *Dir) Delete(ctx context.Context, ref storage.BlobRef) (err error) {
 	// _forwards_, this race should not occur because it is assumed that pieces are never
 	// rewritten with an _older_ storage format version.
 	for i := MinFormatVersionSupported; i <= MaxFormatVersionSupported; i++ {
-		verPath := blobPathForFormatVersion(pathBase, i)
-
-		// move to trash folder, this is allowed for some OS-es
-		moveErr = rename(verPath, trashPath)
-		if os.IsNotExist(moveErr) {
-			// no piece at that path; either it has a different storage format version or there
-			// was a concurrent delete. (this function is expected by callers to return a nil
-			// error in the case of concurrent deletes.)
-			continue
-		}
-		if moveErr != nil {
-			// piece could not be moved into the trash dir; we'll try removing it directly
-			trashPath = verPath
-		}
+		combinedErrors.Add(dir.DeleteWithStorageFormat(ctx, ref, i))
+	}
 
-		// try removing the file
-		err = os.Remove(trashPath)
+	return combinedErrors.Err()
+}
 
-		// ignore concurrent deletes
-		if os.IsNotExist(err) {
-			// something is happening at the same time as this; possibly a concurrent delete,
-			// or possibly a rewrite of the blob. keep checking for more versions.
-			continue
-		}
+// DeleteWithStorageFormat deletes the blob with the specified ref for one specific format version
+func (dir *Dir) DeleteWithStorageFormat(ctx context.Context, ref storage.BlobRef, formatVer storage.FormatVersion) (err error) {
+	defer mon.Task()(&ctx)(&err)
 
-		// the remove may have failed because of an open file handle. put it in a queue to be
-		// retried later.
-		if err != nil {
-			dir.mu.Lock()
-			dir.deleteQueue = append(dir.deleteQueue, trashPath)
-			dir.mu.Unlock()
-		}
+	pathBase, err := dir.blobToBasePath(ref)
+	if err != nil {
+		return err
+	}
+	trashPath := dir.blobToTrashPath(ref)
 
-		// ignore is-busy errors, they are still in the queue
-		// but no need to notify
-		if isBusy(err) {
-			err = nil
-		}
-		combinedErrors.Add(err)
+	verPath := blobPathForFormatVersion(pathBase, formatVer)
+
+	// move to trash folder, this is allowed for some OS-es
+	moveErr := rename(verPath, trashPath)
+	if os.IsNotExist(moveErr) {
+		// no piece at that path; either it has a different storage format
+		// version or there was a concurrent delete. (this function is expected
+		// by callers to return a nil error in the case of concurrent deletes.)
+		return nil
+	}
+	if moveErr != nil {
+		// piece could not be moved into the trash dir; we'll try removing it
+		// directly
+		trashPath = verPath
 	}
 
-	return combinedErrors.Err()
+	// try removing the file
+	err = os.Remove(trashPath)
```

Comment:
I'm confused here. First you move the file to `trashPath` but then you're doing an `os.Remove(trashPath)`? Isn't that deleting the file after moving it?

Reply:
That's exactly right. renaming is an atomic operation in ext3 & 4, so I believe that is why we are first attempting an atomic move, then a delete call.

Reply:
wait, are you sure about ext4?

Reply:
Or FAT32?

Reply:
POSIX requires that rename within filesystem boundaries is atomic.

Reply:
But we don't require POSIX.

Reply:
https://stackoverflow.com/a/41362774

Reply:
You're making a good point, I forget that we don't control our hardware here and we're not as certain about the underlying filesystem. If others want to chime in here that know more, that would be great (they already are).

Another thing to note: this isn't new code, I'm simply refactoring some existing code here to make it usable in another context, so this doesn't introduce any new filesystem semantics.

Reply:
I don't know how realistic this is but I also worry that people might be mounting drives via SAMBA or NFS or other network attached storage.

Reply:
they definitely do/did! @simongui

Reply:
I was responding to the question about ext3/ext4. But yes, os.Rename() will not always be atomic.

Reply:
I've created another ticket to discuss/solve the safety of this part of the code. It wasn't part of this PR and it was existing code so I don't think it should hold these changes back.

Reply:
Where is that ticket? I have some notes for it

Reply:
https://storjlabs.atlassian.net/browse/V3-3037?atlOrigin=eyJpIjoiYzY4OTIwYmVkYjZiNDQ4ODhiNWFjMzdjNTRlYjQ2OGMiLCJwIjoiaiJ9